File: html/js/Tour.js
import Piece from "./Piece.js";

export default class Tour extends Piece
{
    constructor(num, x, y) {super(num, x, y)}

    deplacementValide(xDest, yDest, tabPieces)
    {
        return    super.deplacementValide(xDest, yDest, tabPieces) 
		       && (xDest===this.x || yDest===this.y)
			   && !this.autrePiece(xDest, yDest, tabPieces);
    }

	autrePiece(xDest, yDest, tabPieces)
	{
		let dirX = 1;
		let dirY = 1;
		if(this.x > xDest) dirX = -1;
		if(this.y > yDest) dirY = -1;

		if(yDest === this.y)
			for(let k = 1; k < Math.abs(xDest-this.x); k++)
			{
				if (tabPieces[this.x+k*dirX][this.y] !== null) return true;
			}
		else
			for(let k = 1; k < Math.abs(yDest-this.y); k++)
			{
				if (tabPieces[this.x][this.y+k*dirY] !== null) return true;
			}
		return false;
	}
}

File: java/echec/src/Tour.java
package echec.src;

public class Tour extends Piece
{
	public Tour(int num, int x, int y) {super(num, x, y);}

    public boolean deplacementValide(int xDest,int yDest,Piece[][] tabPieces)
    {
        return    super.deplacementValide(xDest, yDest, tabPieces) 
		       && (xDest==this.x || yDest==this.y)
			   && !this.autrePiece(xDest, yDest, tabPieces);
    }

	private boolean autrePiece(int xDest,int yDest,Piece[][] tabPieces)
	{
		int dirX = 1;
		int dirY = 1;
		if(this.x > xDest) dirX = -1;
		if(this.y > yDest) dirY = -1;

		Piece piece = null;
		if(yDest == this.y)
			for(int k = 1; k < Math.abs(xDest-this.x); k++)
			{
				piece = tabPieces[this.x+k*dirX][this.y];
				if (piece != null && "Roi".equals(piece.getType()) && piece.getCouleur().equals(this.getCouleur()))return false;
				if (piece != null) return true;
			}
		else
			for(int k = 1; k < Math.abs(yDest-this.y); k++)
			{
				piece = tabPieces[this.x][this.y+k*dirY]; 
				if (piece != null && "Roi".equals(piece.getType()) && piece.getCouleur().equals(this.getCouleur()))return false;
				if (piece != null) return true;
			}
		return false;
	}
}
